### PR TITLE
Add Service Principal authentication via CertificateThumprint parameters and correct erroneous GCC High/DOD Defender endpoints

### LINKED ISSUE #63: Explore Replacing the SharePoint PowerShell module with PnP.PowerShell
While looking at M365DSC I noticed that they are not using the official SharePoint PowerShell module. 
They are using the [PnP.PowerShell](https://pnp.github.io/powershell/) module for their SharePoint/OneDrive related resources. 

From a cursory look it seems like it's more fully featured compared to the regular SharePoint PowerShell module and includes Service Principal support that the regular SPO PowerShell module lacks.

### LINKED ISSUE #96: Investigate Using Service principal for Authentication to each of the M365 products
### Overview
Perform a hands-on investigation to determine which of the M365 products (AAD, Defender, etc.) can authenticate using a service principal instead of an interactive user login. The overall goal is to enhance the ScubaGear tool to support service principals for pipelines and other unattended tool execution use cases.

### Scope

- [ ] Create a service principal (application) in Azure AD that will be used as the identity to authenticate to the products

Perform these steps for each of the 7 products that the tool audits (AAD, Defender, EXO, Teams, OneDrive, Sharepoint, PowerPlatform).

- [ ] Review the product documentation for the authentication Cmdlet and determine if it supports client certificates and/or client secrets
- [ ] Setup the certificate or secret in Azure AD
- [ ] Try authenticating to the product using the service principal and document any problems
- [ ] Run the Cmdlets that are used in the respective ScubaGear provider script and make sure they work

- [ ] Create a spreadsheet that contains the details of the investigation

### LINKED ISSUE #144: Defender not Connecting to GCCHigh Correctly
# 🐛 Summary #

When running InvokeScuba for the GCCHigh domain the audits again Defender fails to execute 10 checks because it cannot correctly authenticate into the Security and Compliance center which prohibits it from using Get-DLPCompliancePolicy and Get-DlpComplianceRule cmdlts. 

## To reproduce ##

./InvokeScuba --M365Environment gcchigh

## Expected behavior ##

This should execute all appropriate checks across the different domains. 

## Recommended Fix ##

The ConnectionURI needs to be updated to point to the compliance link and the Authorization Endpoint, which currently defaults to commercial, needs to point to the GCCH endpoint. 

In **ExportDefenderProvider.psm1**
```
            "gcchigh" {
                  # $IPPSParams = $IPPSParams + @{'ConnectionUri' = "https://outlook.office365.us/powershell-liveID";}
		  
		  ##FIXED THE CONNECTION URI - The original URI listed doesn't allow us to the dlp commandlets 
		  $IPPSParams = $IPPSParams + @{'ConnectionUri' = "https://ps.compliance.protection.office365.us/powershell-liveid";}

 		  ##Added the AD Auth Endpoint. Recieve Token Authorization errors if we use the .com auth endpoint. 
		  $IPPSParams = $IPPSParams + @{'AzureADAuthorizationEndpointUri' = "https://login.microsoftonline.us/common";}
            }
```

## Any helpful log output or screenshots ##

```console
Export-DefenderProvider : Error running Connect-IPPSSession. AADSTS900384: JWT token
failed signature validation [Reason - Key was found, but use of the key to verify the
signature failed., Thumbprint of key used by client:
'----', Found key 'Start=12/21/2020 00:00:00,
End=12/21/2025 00:00:00'].
Trace ID: 5cc88641-a8d5-493b-80ef-d94d469b2a00
Correlation ID: 4896588f-7879-49c3-97d2-ccc817bf4807
Timestamp: 2023-02-01 20:12:06Z
At
\path\ScubaGear-0.2.1\PowerShell\ScubaGear\Modules\Orchestrator.psm1:317
char:31
+ ...   $RetVal = Export-DefenderProvider -M365Environment $M365Environment ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Expor
   t-DefenderProvider

WARNING: Omitting the following commands: Get-DlpCompliancePolicy,
Get-DlpComplianceRule, and Get-ProtectionAlert.
```

### PR DESCRIPTION
**Pull request by buidav**
#  #

## 🗣 Description ##

- Add Service Principal Authentication via CertficateThumbprint, AppID, and Organization (domain name).
- Set up `Connection.psm1` for other Service Principal authentication methods to be added in the future.
- Raise minimum versions for EXO and Teams PowerShell modules.
- Add PnP.PowerShell module dependency for Service Principal Authentication support for SPO and OneDrive.
- Handle OneDrive 2.1 which does not have the field we need to check when using PnP.PowerShell.
- Correct erroneous Defender GCCHigh and DOD endpoints.
- Speed up Graph Beta Profile selection for AAD.

## 💭 Motivation and context ##

Closes #96
Closes #63
Closes #144
Future TODO: add other Service Principal Authentication methods.
Future TODO: create a function to register an Azure AD Application with the permissions needed.

## 🧪 Testing ##

 - Ran the updated code in this branch against Test Tenant 4 (E5) with all products individually, mixed, and all products together (*) with no issues with the following command. `Invoke-SCuBA -ProductNames * -CertificateThumbprint $CertThumbprint -AppID $AppID -Organization "tenant4.onmicrosoft.com"`
 - Compared results of the tests above with copies run from interactive login via `Invoke-SCuBA -ProductNames *`. No regressions found for E5 tenants.
 - [See rough notes in this comment for how to create your own application for testing](https://github.com/cisagov/ScubaGear/issues/96#issuecomment-1413129118)

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] ~~I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.~~
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

### CODE REVIEW

File: PowerShell/ScubaGear/Modules/Providers/ExportOneDriveProvider.psm1 [hunk 1]
```diff
@@ -22,21 +26,34 @@ function Export-OneDriveProvider {
     $InitialDomain = ($Tracker.TryCommand("Get-MgOrganization")).VerifiedDomains | Where-Object {$_.isInitial}
```
Comment (Sloane4) [inline PowerShell/ScubaGear/Modules/Providers/ExportOneDriveProvider.psm1:26]: Can be removed as obsolete code

File: PowerShell/ScubaGear/Modules/Providers/ExportOneDriveProvider.psm1 [hunk 2]
```diff
@@ -22,21 +26,34 @@ function Export-OneDriveProvider {
...
     $InitialDomainPrefix = $InitialDomain.Name.split(".")[0]
```
Comment (Sloane4) [inline PowerShell/ScubaGear/Modules/Providers/ExportOneDriveProvider.psm1:27]: Can be removed as obsolete code

File: PowerShell/ScubaGear/Modules/Providers/ExportOneDriveProvider.psm1 [hunk 3]
```diff
@@ -22,21 +26,34 @@ function Export-OneDriveProvider {
...
 
-    #Get SPOSiteIdentity
     $SPOSiteIdentity = Get-SPOSiteHelper -M365Environment $M365Environment -InitialDomainPrefix $InitialDomainPrefix
```
Comment (Sloane4) [inline PowerShell/ScubaGear/Modules/Providers/ExportOneDriveProvider.psm1:29]: Can be removed as obsolete code

File: PowerShell/ScubaGear/Modules/Providers/ExportOneDriveProvider.psm1 [hunk 4]
```diff
@@ -22,21 +26,34 @@ function Export-OneDriveProvider {
...
+    Write-Verbose "This is the tenant's main sharepoint domain use this for -Identity calls: $($SPOSiteIdentity)"
```
Comment (Sloane4) [inline PowerShell/ScubaGear/Modules/Providers/ExportOneDriveProvider.psm1:30]: Not necessary, remove.

File: PowerShell/ScubaGear/Modules/Providers/ExportOneDriveProvider.psm1 [hunk 5]
```diff
@@ -22,21 +26,34 @@ function Export-OneDriveProvider {
...
 
     $SuccessfulCommands = ConvertTo-Json @($Tracker.GetSuccessfulCommands())
     $UnSuccessfulCommands = ConvertTo-Json @($Tracker.GetUnSuccessfulCommands())
 
     # Note the spacing and the last comma in the json is important
     $json = @"
     "SPO_tenant_info": $SPOTenantInfo,
-    "Expected_results": $ExpectedResults,
     "Tenant_sync_info": $TenantSyncInfo,
+    "OD_used_PnP": $UsedPnp,
```
Comment (Sloane4) [inline PowerShell/ScubaGear/Modules/Providers/ExportOneDriveProvider.psm1:56]: Refactor with a better name

**Review by Sloane4 — APPROVED** [review]

**Comment by tkol2022** [comment]
### Testing Update

**AAD**

- Tested the execution of AAD with a service principal and it generates a report without any errors.
- Ran the branch against a golden copy of the AAD json files using RunFunctionalTests and it passes. No changes to the policy outputs were detected.
- Ran ScubaGear using the old interactive login and it still functions as expected.
- We verified that it is NOT necessary to assign a role to the service principal for it to work.

**Review by tkol2022 — APPROVED** [review]

**Comment by ethanb-cisa** [comment]
@tkol2022 what appId did you use in the G5 tenant to test against EXO. I want to check something.

**Comment by nanda-katikaneni** [comment]
> @tkol2022 what appId did you use in the G5 tenant to test against EXO. I want to check something.
Here is the info:
Name: Teds Service Principal
App ID: b682d3a5-bc68-450e-80ba-018b97aa2b21

**Pull request merged**